### PR TITLE
[Bug] Fixed progression switching is inconsistent with camera rails: switch character in the middle of rail transition

### PR DESCRIPTION
**Pull request by h4rdmol**
### [Bug] Progression switching is inconsistent with camera rails like when moving to the previous character - Switch character in NMM module only on half of rail or as usual with `Instant Character Switch` enabled

https://trello.com/c/HuRuz2rh/639-5-progression-system

**Issus Summary:** 

In MainMenuWidget the SwitchCurrentPlayer was changing the new spot mesh component on the camera rail start which caused the PlayerTypeChanged event invoked faster than the actual camera rig transition completed. It was completely ignoring the camera rail state.

**Solution:** 

I've introduced CameraRailTransitionState events. So now the MainMenuWidget sets the next Spot's mesh on the Camera rail transition on halfway of its path. The transition now from unlocked to locked level will be initiated in the middle of the camera rail. 

Reverted data asset:
****image

### CODE REVIEW

File: Content/Bomber/DataAssets/DA_GameState.uasset
Comment (JanSeliv) [inline Content/Bomber/DataAssets/DA_GameState.uasset:1]: Please revert DA_GameState as it seems unrelated

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Subsystems/NMMCameraSubsystem.cpp [hunk 1]
```diff
@@ -244,11 +255,38 @@ void UNMMCameraSubsystem::TickTransition(float DeltaTime)
...
 	if (FMath::IsNearlyEqual(Progress, GetCameraLastTransitionValue(), KINDA_SMALL_NUMBER))
 	{
 		OnEndTransition();
 	}
 }
+
+// Is called to check if camera rail is halfway to another spot
+void UNMMCameraSubsystem::IsHalfwayTransition()
+{
+	ACineCameraRigRail* CurrentRailRig = GetCurrentRailRig();
```
Comment (JanSeliv) [inline Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Subsystems/NMMCameraSubsystem.cpp:274]: CurrentRailRig can be null, wrap it by ensure

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Subsystems/NMMCameraSubsystem.cpp [hunk 2]
```diff
@@ -244,11 +255,38 @@ void UNMMCameraSubsystem::TickTransition(float DeltaTime)
...
+
+	// Check if halfway reached
+	if (IsCameraForwardTransition())
```
Comment (JanSeliv) [inline Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Subsystems/NMMCameraSubsystem.cpp:277]: - I believe the whole logic of IsHalfwayTransition function can be twice smaller: you don't need to know to which direction camera transits (forward or backward), but just check is it center now, like next:
```
constexpr float HalfwayPosition = 0.5f;
if (CameraRailTransitionStateInternal != ENMMCameraRailTransitionState::HalfwayTransition && 
    FMath::IsNearlyEqual((Progress, HalfwayPosition, KINDA_SMALL_NUMBER))
{
    SetNewCameraRailTranitionState(ENMMCameraRailTransitionState::HalfwayTransition);
}
```
- Please move this entire logic back to `TickTransition` as you see it's just one if-condition

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp [hunk 3]
```diff
@@ -17,6 +17,8 @@
 #include "Components/Button.h"
 #include "Kismet/KismetSystemLibrary.h"
 //---
+#include "Subsystems/NMMCameraSubsystem.h"
```
Comment (JanSeliv) [inline Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp:20]: Please take this include higher to be under first category (first `//---`), in the area where `include "Bomber.h"`

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Subsystems/NMMCameraSubsystem.h [hunk 4]
```diff
@@ -17,6 +18,14 @@ class NEWMAINMENU_API UNMMCameraSubsystem : public UTickableWorldSubsystem
 	GENERATED_BODY()
 
 public:
+	DECLARE_DYNAMIC_MULTICAST_DELEGATE_OneParam(FNMMOnCameraRailTransitionStateChange, ENMMCameraRailTransitionState, CameraRailState);
+	
+	/** Called when the Camera r
+	 * ail End Transition is called
```
Comment (JanSeliv) [inline Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Subsystems/NMMCameraSubsystem.h:24]: Something wrong with description

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Subsystems/NMMCameraSubsystem.h [hunk 5]
```diff
@@ -73,11 +82,28 @@ class NEWMAINMENU_API UNMMCameraSubsystem : public UTickableWorldSubsystem
...
+	FORCEINLINE ENMMCameraRailTransitionState GetCurrentCameraRailTransitionState() const { return CameraRailTransitionStateInternal; }
+
 protected:
 	/** Is true during the transition when the camera is currently blending in to the rail or out from the Rail to the Spot. */
 	UPROPERTY(VisibleInstanceOnly, BlueprintReadWrite, Transient, Category = "C++", meta = (BlueprintProtected, DisplayName = "Is Blending In/Out"))
 	bool bIsBlendingInOutInternal = false;
 
+	/** Is true when the rail of the camera is currently halfway of the current Spot to the next */
+	UPROPERTY(VisibleInstanceOnly, BlueprintReadWrite, Transient, Category = "C++", meta = (BlueprintProtected, DisplayName = " Is Rail Half Way Reached"))
+	bool bIsRailHalfWayReached = false;
```
Comment (JanSeliv) [inline Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Subsystems/NMMCameraSubsystem.h:101]: This `bIsRailHalfWayReached` property seems to be not used: you just broadcast delegate instead of relying on property. Please remove unused property

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Widgets/NewMainMenuWidget.h [hunk 6]
```diff
@@ -46,6 +47,10 @@ class NEWMAINMENU_API UNewMainMenuWidget : public UUserWidget
 	UPROPERTY(EditDefaultsOnly, BlueprintReadWrite, Transient, Category = "C++", meta = (BlueprintProtected, BindWidget))
 	TObjectPtr<UButton> QuitGameButton = nullptr;
 
+	/** The current Main Menu Spot */
+	UPROPERTY(VisibleInstanceOnly, BlueprintReadWrite, Transient, Category = "C++", meta = (BlueprintProtected))
+	TObjectPtr<UNMMSpotComponent> MainMenuSpotInternal = nullptr;
```
Comment (JanSeliv) [inline Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Widgets/NewMainMenuWidget.h:52]: You don't need to cache any spots, you always can get current one from `UNMMSpotsSubsystem::Get().GetCurrentSpot()`

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Widgets/NewMainMenuWidget.h [hunk 7]
```diff
@@ -4,6 +4,7 @@
 
 #include "Blueprint/UserWidget.h"
 //---
+#include "Components/NMMSpotComponent.h"
```
Comment (JanSeliv) [inline Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Widgets/NewMainMenuWidget.h:7]: Try always avoid Includes in headers. This one can definitely be removed

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp [hunk 8]
```diff
@@ -193,3 +211,12 @@ void UNewMainMenuWidget::OnQuitGameButtonPressed()
 	AMyPlayerController* MyPC = GetOwningPlayer<AMyPlayerController>();
 	UKismetSystemLibrary::QuitGame(this, MyPC, EQuitPreference::Background, false);
 }
+
+// Called when the Camera Rail transition state changed
+void UNewMainMenuWidget::OnCameraRailTransitionStateChanged_Implementation(ENMMCameraRailTransitionState CameraRailTransitionStateChanged)
+{
+	if (CameraRailTransitionStateChanged == ENMMCameraRailTransitionState::HalfwayTransition)
+	{
+		UpdatePlayerCharacterMesh(MainMenuSpotInternal ? MainMenuSpotInternal->GetMeshChecked().GetCustomPlayerMeshData() : FCustomPlayerMeshData::Empty);
```
Comment (JanSeliv) [inline Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp:220]: Please never put ternary into any function argument, but make it as local variable instead, like next:
```
const FCustomPlayerMeshData& PlayerMeshData = MainMenuSpotInternal ? MainMenuSpotInternal->GetMeshChecked().GetCustomPlayerMeshData() : FCustomPlayerMeshData::Empty;
UpdatePlayerCharacterMesh(PlayerMeshData);
```

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Subsystems/NMMCameraSubsystem.cpp (hunk 2 above)
Comment (h4rdmol) [reply Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Subsystems/NMMCameraSubsystem.cpp]: done as part of https://github.com/JanSeliv/Bomber/pull/70/commits/e9901bd0f1418c25167358664b233f58412be169

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp (hunk 3 above)
Comment (h4rdmol) [reply Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp]: done as part of https://github.com/JanSeliv/Bomber/pull/70/commits/c7b3f517ae62711dbaf14c873bcc0489cdc2e2cf

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Subsystems/NMMCameraSubsystem.h (hunk 4 above)
Comment (h4rdmol) [reply Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Subsystems/NMMCameraSubsystem.h]: fixed typo as part of https://github.com/JanSeliv/Bomber/pull/70/commits/c7b3f517ae62711dbaf14c873bcc0489cdc2e2cf

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Subsystems/NMMCameraSubsystem.h (hunk 5 above)
Comment (h4rdmol) [reply Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Subsystems/NMMCameraSubsystem.h]: removed https://github.com/JanSeliv/Bomber/pull/70/commits/7f7bfda645dac70304e5c0f73c9c809ef9b0ba9c

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Widgets/NewMainMenuWidget.h (hunk 6 above)
Comment (h4rdmol) [reply Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Widgets/NewMainMenuWidget.h]: removed https://github.com/JanSeliv/Bomber/pull/70/commits/9dbca39446b11b4ce0681f75d994c774033abb51

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Widgets/NewMainMenuWidget.h (hunk 7 above)
Comment (h4rdmol) [reply Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Widgets/NewMainMenuWidget.h]: cleaned header https://github.com/JanSeliv/Bomber/pull/70/commits/9dbca39446b11b4ce0681f75d994c774033abb51

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp (hunk 8 above)
Comment (h4rdmol) [reply Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp]: got it, used local variable instead https://github.com/JanSeliv/Bomber/pull/70/commits/9dbca39446b11b4ce0681f75d994c774033abb51

File: Content/Bomber/DataAssets/DA_GameState.uasset
Comment (h4rdmol) [reply Content/Bomber/DataAssets/DA_GameState.uasset:1]: image
DA_GameState was reverted to initial state as part of the https://github.com/JanSeliv/Bomber/pull/70/commits/48b64b549ae5f94f348d85b71faee3d0d8d79661

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp (hunk 8 above)
Comment (JanSeliv) [reply Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp]: @h4rdmol is not finished yet.
- You copy PlayerMeshData while on my example I used const ref: `const FCustomPlayerMeshData& PlayerMeshData = ...`
- You don't need to call `UNMMSpotsSubsystem::Get().GetCurrentSpot()` here, since few lines above `UNMMSpotsSubsystem::Get().MoveMainMenuSpot(Incrementer);` already can return you the spot.

Please do like next, look at code carefully:
```
	const UNMMSpotComponent* MenuSpot = UNMMSpotsSubsystem::Get().MoveMainMenuSpot(Incrementer);
	if (UNMMBaseSubsystem::Get().GetCurrentMenuState() == ENMMState::Idle)
	{
		const FCustomPlayerMeshData& PlayerMeshData = MenuSpot ? MenuSpot->GetMeshChecked().GetCustomPlayerMeshData() : FCustomPlayerMeshData::Empty;
		UpdatePlayerCharacterMesh(PlayerMeshData);
	}
```

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp [hunk 9]
```diff
@@ -13,10 +13,13 @@
 #include "Subsystems/SoundsSubsystem.h"
 #include "UI/SettingsWidget.h"
 #include "UtilityLibraries/MyBlueprintFunctionLibrary.h"
+#include "Subsystems/NMMCameraSubsystem.h"
 //---
 #include "Components/Button.h"
 #include "Kismet/KismetSystemLibrary.h"
 //---
+
+
```
Comment (JanSeliv) [inline Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp:22]: Please remove your extra spaces from review. There are few of them, please go through your changes

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp (hunk 8 above)
Comment (h4rdmol) [reply Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp]: got it https://github.com/JanSeliv/Bomber/pull/70/commits/a6bc8187296a0edd727fd7ec33cfeff6fe42689d

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp (hunk 9 above)
Comment (h4rdmol) [reply Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp]: done https://github.com/JanSeliv/Bomber/pull/70/commits/078e168f379072ecd945a7d7da4116fba448108e

Comment (JanSeliv) [reply Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp]: @h4rdmol is not finished yet
image

Comment (JanSeliv) [reply Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp]: image

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Subsystems/NMMCameraSubsystem.cpp [hunk 10]
```diff
@@ -186,6 +186,13 @@ float UNMMCameraSubsystem::GetCameraLastTransitionValue()
 	return IsCameraForwardTransition() ? 1.f : 0.f;
 }
 
+// Applies the new state of camera rail transition state
+void UNMMCameraSubsystem::SetNewCameraRailTransitionState(ENMMCameraRailTransitionState NewCameraRailState)
+{
+	CameraRailTransitionStateInternal = NewCameraRailState;
```
Comment (JanSeliv) [inline Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Subsystems/NMMCameraSubsystem.cpp:192]: please make sure first state is not the same just for safety, so system guarantees we will not spamming broadcast. Like next:

```
	if (NewCameraRailState == CameraRailTransitionStateInternal)
	{
		return;
	}

	CameraRailTransitionStateInternal = NewCameraRailState;
	OnCameraRailTransitionStateChanged.Broadcast(CameraRailTransitionStateInternal);
```

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Data/NMMTypes.h [hunk 11]
```diff
@@ -61,3 +61,19 @@ enum class ENMMState : uint8
...
+UENUM(BlueprintType, DisplayName = "New Main Menu Camera Rail State")
+enum class ENMMCameraRailTransitionState : uint8
+{
+	///< Transition is not started
+	None,
+	///< Camera moving between Spots started
+	BeginTransition,
+	///< Camera moving between Spots reached it's halfway
+	HalfwayTransition,
+	///< Camera moving between Spots finished tranisition
```
Comment (JanSeliv) [inline Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Data/NMMTypes.h:77]: `tranisition` typo

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Subsystems/NMMCameraSubsystem.cpp (hunk 10 above)
Comment (h4rdmol) [reply Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Subsystems/NMMCameraSubsystem.cpp:197]: addressed https://github.com/JanSeliv/Bomber/pull/70/commits/c22b3990b919374d1a74ae847b0f8177c94bf4ae

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp (hunk 9 above)
Comment (h4rdmol) [reply Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Private/Widgets/NewMainMenuWidget.cpp]: resolved as part of: https://github.com/JanSeliv/Bomber/pull/70/commits/05d289d60373fbdce4513111385d8dfbd29138ac and https://github.com/JanSeliv/Bomber/pull/70/commits/d458ad88bcba1ad65c8e50407f2ab9990332aa6c

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Widgets/NewMainMenuWidget.h [hunk 12]
```diff
@@ -75,6 +75,11 @@ class NEWMAINMENU_API UNewMainMenuWidget : public UUserWidget
 	UFUNCTION(BlueprintCallable, Category = "C++")
 	void SwitchCurrentPlayer(int32 Incrementer);
 
+	/** Update the chosen player mesh on the level
+	* @param CustomPlayerMeshData is the new player mesh */
+	UFUNCTION(BlueprintCallable, Category = "C++")
+	void UpdatePlayerCharacterMesh(const FCustomPlayerMeshData& CustomPlayerMeshData);
```
Comment (JanSeliv) [inline Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Widgets/NewMainMenuWidget.h:81]: Did you setup non-unity build? I feel your changes might fail since you did not preforward `FCustomPlayerMeshData` and `ENMMCameraRailTransitionState`.

- Please add forwards on the beginning:
```
struct FCustomPlayerMeshData;
enum class ENMMCameraRailTransitionState : uint8;
```
- Locally compile NewMainMenu.Build.cs with non-unity and without PCHUsage

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Data/NMMTypes.h (hunk 11 above)
Comment (h4rdmol) [reply Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Data/NMMTypes.h]: fixed https://github.com/JanSeliv/Bomber/pull/70/commits/35e60062c443c7a66e00f6bf46827a3dc18671d8

File: Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Widgets/NewMainMenuWidget.h (hunk 12 above)
Comment (h4rdmol) [reply Plugins/GameFeatures/NewMainMenu/Source/NewMainMenu/Public/Widgets/NewMainMenuWidget.h:83]: done, no issues detected. Added forwarding: https://github.com/JanSeliv/Bomber/pull/70/commits/05cb2cecad50fcf33f231deeaedaff46df415a2f
